aspnet/mvc/controllers/routing.rst

```diff
@@ -1,8 +1,140 @@
+
+Routing is part of ASP.NET Core middleware. Middleware is a component that is assembled into an application pipeline to handle requests and responses. It's the middleware that processes the request until it gets to routing, which then hands over the request to MVC. MVC then selects an action method based on data that comes from routing. After that point it's up to the developer to add code to the action method to retrieve and modify data, or call services. 
+
+Startup code and relationship to MVC route handler
+--------------------------------------------------
+
+To configure routing, add the following code to the ``Configure`` method in the ``Startup.cs`` file. The code maps a route named ``default`` to a controller named ``Home`` and action method named ``Index``. You configure this by indicating which controllers, action methods, and parameters you want recognized as a route in the ``template`` argument of the ``MapRoute`` method, as shown below: 
+
+.. code-block:: c# 
+
+  app.UseMvc(routes =>
+  {
+    routes.MapRoute(
+      name: "default",
+      template: "{controller=Home}/{action=Index}/{id?}");
+  });
+  
+The route template defines the route as well as default values, in this example, ``Home`` for the controller, ``Index`` for the action method, and optionally ``id`` as a URL parameter. Routes defined in the ``Startup.cs`` file are conventional routes. 
+
+Conventional routing
+--------------------
+
+Routing first matches the URL to a route template, then selects a controller and action to route the request. You may define routes and route constraints on both controllers and actions. In conventional routing, you define a route that generates the action and controller tokens. The value of those tokens are used to select an action and controller by name. For example, the URL ending in ``/home/index/`` matches the route template for ``{controller}/{action}``. Conventional routing must be configured in the ``Startup`` class, as shown in the previous code sample. 
+
+For most forms-over-data web applications that use standard controllers, conventional routes work nicely. Attribute routing is the recommended type of routing for use in APIs.
+
+Attribute routing
+-----------------
+
+[Route],[Http*]
+^^^^^^^^^^^^^^^
+
+Rather having to rely on naming and conventions to figre out where to route the request, attribute routing enables you to use an attribute to route requests directly to the desired destination. You may apply attributes to either controllers or actions. Attribute routing is more direct because routes must match what is defined in the attribute in order to be processed. When an action method is decorated with the ``[Route]`` attribute, it can no longer be accessed through the convention-based routes, as the attribute takes precedence. 
```

Comment:
> Attribute routing is more direct because routes must match what is defined in the attribute in order to be processed

"URLs must match what is defined in the attribute..."